docs/design/values.md

```diff
@@ -623,6 +626,121 @@ The model of initialization of returns also facilitates the use of
 [`returned var` declarations](control_flow/return.md#returned-var). These
 directly observe the storage provided for initialization of a function's return.
 
+### Expression forms
+
+We typically treat the category and type of an expression as independent
+properties. However, in some cases they can become entangled, and we need deal
+with them as an integrated whole. The _form_ of an expression captures all of
+the information about it that is visible to the type system, while abstracting
+away all other information about it. Thus, forms are a generalization of types:
+what we conventionally call "types" are really the types of objects and values,
+whereas forms are the types of expressions and patterns.
+
+A _primitive form_ consists of a type, an expression category, an expression
+phase, and optionally a constant value (which is present if and only if the
+expression phase is not "runtime"). When dealing with primitive forms, which is
+the common case, we can treat each of those properties as independent. For
+convenience, in this section we will use the notation `[T, C, P, V]` to
+represent a primitive form with type `T`, category `C`, phase `P` and value `V`,
+but this is not Carbon syntax.
+
+Other forms are called _composite forms_, and there are two kinds:
+
+A _tuple form_ can be thought of as a tuple of forms, just as a tuple type can
+be thought of as a tuple of types. The form of a tuple literal is a tuple form,
+whose elements are the forms of the literal elements.
+
+> **TODO:** Extend this to support variadic forms.
+
+A _struct form_ can be thought of as a struct whose fields are forms, just as a
+struct type can be thought of as a struct whose fields are types. The form of a
+struct literal is a struct form with the same field names, whose values are the
+forms of the corresponding fields of the struct literal.
+
+An expression with a composite form still has a well-defined type, phase, and
+(if the phase is not "runtime") constant value. However, it does not have a
+well-defined expression category.
+
+#### Form conversions
+
+An expression with a primitive form can be converted to another primitive form
+by applying type, category, and phase conversions independently. Form
```

Comment:
I don't think these are independent. For example, to go from a value to initializing a `var` pattern requires that the type be copyable (or an implicit conversion exists between the source and destination types).